hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/HazelcastCaseOperator.java

```diff
@@ -0,0 +1,201 @@
+
+public final class HazelcastCaseOperator extends SqlOperator {
+
+    public static final HazelcastCaseOperator INSTANCE = new HazelcastCaseOperator();
+
+    private HazelcastCaseOperator() {
+        super(SqlCaseOperator.INSTANCE.getName(), SqlKind.CASE, SqlCaseOperator.INSTANCE.getLeftPrec(), true,
+                wrap(new CaseReturnTypeInference()), null, null);
+    }
+
+    @Override
+    public void validateCall(SqlCall call, SqlValidator validator, SqlValidatorScope scope, SqlValidatorScope operandScope) {
+        SqlCaseOperator.INSTANCE.validateCall(call, validator, scope, operandScope);
+    }
+
+    @Override
+    public RelDataType deriveType(SqlValidator validator, SqlValidatorScope scope, SqlCall call) {
+        // SqlCaseOperator is doing the same
+        return validateOperands(validator, scope, call);
+    }
+
+    @Override
+    // override this methods because passing null into constructor for SqlOperandTypeChecker
+    public SqlOperandCountRange getOperandCountRange() {
+        return SqlOperandCountRanges.any();
+    }
+
+    @Override
+    public boolean checkOperandTypes(SqlCallBinding callBinding, boolean throwOnFailure) {
+        return SqlCaseOperator.INSTANCE.checkOperandTypes(callBinding, throwOnFailure);
+    }
+
+    @Override
+    public SqlSyntax getSyntax() {
+        return SqlCaseOperator.INSTANCE.getSyntax();
+    }
+
+    @Override
+    public void unparse(SqlWriter writer, SqlCall call, int leftPrec, int rightPrec) {
+        assert call.getOperandList().size() == 3;
+
+        final SqlWriter.Frame frame = writer.startList(SqlWriter.FrameTypeEnum.CASE, "CASE", "END");
+
+        SqlNodeList whenList = (SqlNodeList) call.getOperandList().get(0);
+        SqlNodeList thenList = (SqlNodeList) call.getOperandList().get(1);
+        assert whenList.size() == thenList.size();
+        for (Pair<SqlNode, SqlNode> pair : Pair.zip(whenList, thenList)) {
+            writer.sep("WHEN");
+            pair.left.unparse(writer, 0, 0);
+            writer.sep("THEN");
+            pair.right.unparse(writer, 0, 0);
+        }
+
+        writer.sep("ELSE");
+        SqlNode elseExpr = call.getOperandList().get(2);
+        elseExpr.unparse(writer, 0, 0);
+        writer.endList(frame);
+    }
+
+    @Override
+    public SqlCall createCall(SqlLiteral functionQualifier, SqlParserPos pos, SqlNode... operands) {
+        return new HazelcastSqlCase(pos, operands[0], (SqlNodeList) operands[1], (SqlNodeList) operands[2], operands[3]);
+    }
+
+    private static class CaseReturnTypeInference implements SqlReturnTypeInference {
+        @Override
+        @SuppressWarnings({"checkstyle:npathcomplexity", "checkstyle:cyclomaticcomplexity", "checkstyle:nestedifdepth"})
+        public RelDataType inferReturnType(SqlOperatorBinding binding) {
+            // Copied from SqlCaseOperator#inferTypeFromValidator with small changes
```

Comment:
What are those changes? Maybe a short summary of what and why would be useful as this method is quite complex...